File: Concurrency/Java/fizz-buzz-multithreaded.java
class FizzBuzz {
    private int n;
    private Semaphore fizz;
    private Semaphore buzz;
    private Semaphore fizzbuzz;
    private Semaphore number;
    private int it = 1;

    public FizzBuzz(int n) {
        this.n = n;
        this.fizz = new Semaphore(0);
        this.buzz = new Semaphore(0);
        this.fizzbuzz = new Semaphore(0);
        this.number = new Semaphore(0);
    }

    // printFizz.run() outputs "fizz".
    public void fizz(Runnable printFizz) throws InterruptedException {
        while(true)
        {
            fizz.acquire();
            if(it>n)break;
            printFizz.run();
            number.release();
        }
    }

    // printBuzz.run() outputs "buzz".
    public void buzz(Runnable printBuzz) throws InterruptedException {
        while(true)
        {
            buzz.acquire();
            if(it>n)break;
            printBuzz.run();
            number.release();   
        }
    }

    // printFizzBuzz.run() outputs "fizzbuzz".
    public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        while(true)
        {
            fizzbuzz.acquire();
            if(it>n)break;
            printFizzBuzz.run();
            number.release();   
        }
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public void number(IntConsumer printNumber) throws InterruptedException {
        for(it=1;it<=n;it++)
        {
            if(it%3 == 0 && it%5 == 0)
            {
                fizzbuzz.release();
                number.acquire();
            }else if(it%3 == 0)
            {
                fizz.release();
                number.acquire();
            }else if(it%5 == 0)
            {
                buzz.release();
                number.acquire();
            }else{
                printNumber.accept(it);   
            }
        }
        fizz.release();
        buzz.release();
        fizzbuzz.release();
    }
}

File: Concurrency/C#/fizz-buzz-multithreaded.cs
using System.Threading;
public class FizzBuzz {
    private int n;
    private EventWaitHandle fizz;
    private EventWaitHandle buzz;
    private EventWaitHandle fizzbuzz;
    private EventWaitHandle number;
    private int it = 1;
    public FizzBuzz(int n) {
        this.n = n;
        this.fizz = new AutoResetEvent(initialState: false);
        this.buzz = new AutoResetEvent(initialState: false);
        this.fizzbuzz = new AutoResetEvent(initialState: false);
        this.number = new AutoResetEvent(initialState: true);
    }

    // printFizz() outputs "fizz".
    public void Fizz(Action printFizz) {
        while(true)
        {
            fizz.WaitOne();
            if(it>n)break;
            printFizz();
            number.Set();
        }
    }

    // printBuzzz() outputs "buzz".
    public void Buzz(Action printBuzz) {
        while(true)
        {
            buzz.WaitOne();
            if(it>n)break;
            printBuzz();
            number.Set();   
        }
    }

    // printFizzBuzz() outputs "fizzbuzz".
    public void Fizzbuzz(Action printFizzBuzz) {
        while(true)
        {
            fizzbuzz.WaitOne();
            if(it>n)break;
            printFizzBuzz();
            number.Set();   
        }
    }

    // printNumber(x) outputs "x", where x is an integer.
    public void Number(Action<int> printNumber) {
        for(it=1;it<=n;it++)
        {
            if(it%3 == 0 && it%5 == 0)
            {
                fizzbuzz.Set();
                number.WaitOne();
            }else if(it%3 == 0)
            {
                fizz.Set();
                number.WaitOne();
            }else if(it%5 == 0)
            {
                buzz.Set();
                number.WaitOne();
            }else{
                printNumber(it);   
            }
        }
        fizz.Set();
        buzz.Set();
        fizzbuzz.Set();
    }
}
